tar_scm.py

```diff
@@ -72,6 +72,17 @@ def fetch_upstream_git(url, clone_dir, revision, cwd, kwargs):
     """Fetch sources via git."""
     safe_run(['git', 'clone', url, clone_dir], cwd=cwd,
              interactive=sys.stdout.isatty())
+    if revision:
+        # check if the reference already exists.
+        try:
+            safe_run(['git', 'rev-parse', '--verify', '--quiet', revision],
```

Comment:
I'm not sure I follow you.  Why would we want to change the revision to `'origin/' + revision`?  My point was that the following sequence contains redundancy:

1. `git clone`
2. Check if we have the revision we want.
3. If not fetch it.
4. Check if we have the revision we want.
5. Switch to it.

But I suppose it could be argued that step 4 is an extra safeguard against step 3. not doing what we expect.  So I'm OK with keeping that, as long as the duplicate code for steps 2. and 4. is refactored into a separate function like I already suggested to do.